WinInstaller/WinInstaller/Transaction.h

```diff
@@ -0,0 +1,114 @@
+#pragma once
+#include <functional>
+#include <stack>
+#include <iostream>
+#include <string>
+
+
+namespace mywininstaller
+{
+	template <typename T>
+	class Transaction
+	{
+	public:
+		Transaction() = default;
+		~Transaction();
+		Transaction(const Transaction<T>& other) = delete;
+		Transaction(Transaction<T>&& other) = delete;
+		Transaction<T>& operator=(const Transaction<T>& other) = delete;
+		Transaction<T>& operator=(Transaction<T>&& other) = delete;
+		void rollback(bool throwOnError = true);
+		void commit();
+		void addAction(const T& object, std::function<void(T)> rollbackFunc, const std::string& name = "");
```

Comment:
what is the expected behaviour of an "action" here?

1. exposing a rollback(true) api indicates you expect the rollback function to be callable more than once
in what state should object remain in this case?
exposing such a feature imposes constraints of the implementor of rollbackFunc, but you didn't document this point.
reconsider if you even need this feature

2. who manages the lifetime of object? current code suggest object must be copyable. A very limiting design choice, what if object is stateful (e.g. holds logical resources) and cannot be copied?

3. separating object from rollback introduces this template and a very non- OOP code
the action itself is separated from the rollback, which are very much logically linked. It would be better if both are managed by the same entity. 

Take these 3 suggestions and simplify the design

Reply:
**About point 1:**
The idea of exposing `rollback` as public was to allow anyone who uses this class to catch exceptions thrown during rollback, instead of always swallowing them. 
`rollbackFunc` accepts object by value, doesn't that guarantee that it will remain with valid state if anyone calls `rollback` again? (the `Transaction` object doesn't remove it from the stack if an exception is thrown)

Anyway those are good points that reveal the design isn't good enough. See below for a new suggestion.

**New design:**

- We introduce an abstract base class `TransactedAction` with a virtual functions `execute` and `rollback`.
Anyone that wants an action that can be undone can subclass this.

- `Transaction` object holds a stack of `TransactedAction`s. It calls `execute` on add, and `rollback` on rollback.

- `TransactedAction`s are **moved** when handed to `Transaction`, so that all lifetime management happens in the `Transaction` object from then on.

- This makes the template redundant. `Transaction` can be a normal class.

- It also removes the constraint that all rollbacks must act on the same object type, since we have a polymorphic base class with a known interface.


I'm still undecided about wether to leave `Transaction::rollback()` as public.
I think it's nice to have the option to catch exceptions in rollback, handle them and then call rollback again and continue from the action that failed. Do you think there are major downsides to this now that we have the new design?